lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,231 @@
+require 'set'
+require 'yaml'
+require 'open-uri'
+require 'objspace'
+require 'fog/core'
+
+module Fog
+  module Orchestration
+    module Util
+      #
+      # Resolve get_file resources found in a HOT template populating
+      #  a files Hash conforming to Heat Specs
+      #  https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=create-stack-detail#stacks
+      #
+      # This implementation just process nested templates but not resource
+      #  registries.
+      class RecursiveHotFileLoader
+        attr_reader :files
+        attr_reader :template
+        attr_reader :visited
+
+        def initialize(template, files = nil)
+          # Serialize the template hash to deep-copy it and
+          #  avoid modifying the argument. Instead create a
+          #  new one to be modified by get_file_contents.
+          @template = deep_copy(template)
+          @files = files || {}
+          @visited = Set.new
+        end
+
+        def get_files
+          return @files unless @files.empty?
```

Comment:
```
return @files unless @files.nil? # avoid processing templates for every `get_files` call
@files = {}
```

Reply:
@bzwei agreed on gitter that we'll reintroduce `files` argument.